### ISSUE TITLE
Made with CNL 

### ISSUE THREAD

**Issue by pgruenbacher** [johnmcfarlane/cnl#323]
I think it'd be useful to have a list of projects (if we know of any) that are using the library to handle their numeric calculations. Likewise it may be worth adding the cppcon youtube videos of CNL onto the README as an intro.

**Comment by johnmcfarlane** [comment]
Oh you mean embedded? I hadn't thought of that. Generally I need to do a major sweep of the docs. Thanks

**Comment by pgruenbacher** [comment]
I'll add to this issue as I try integrating this library. I think other stuff missing from the docs is the multiply api has change, there's safe_multiply now as well, etc.

I do like the saturated_overflow and the multiply(tag idea. It allows me to develop with the exception tag and then switch to deterministic saturated tag in production.

**Comment by pgruenbacher** [comment]
ok so I kept getting confused by the docs / presentations / current code.
It sounds like the explicit arithmetic operators ("multiply", "divide") got dropped from the proposal.
And so it seems at the moment, if I want to implement fixed point math, is I should do something like 
```
template<
        int IntegerDigits,
        int FractionalDigits,
        class Narrowest>
using saturated_elastic_fixed_point = cnl::fixed_point<
        cnl::overflow_integer<
                cnl::elastic_integer<
                        IntegerDigits+FractionalDigits,
                        Narrowest>,
                cnl::saturated_overflow_tag>,
        -FractionalDigits>;
using real_t3 = saturated_elastic_fixed_point<7, 8, std::int16_t>;
```
and then just treat it as if were a real float type. I can change the tag from saturated to throwing as I develop, and it seems this will just play well with std::sqrt too

**Comment by pgruenbacher** [comment]
I don't see a divide function available in the api besides divide, but divide only seems to work on ints. It seems to work on fractions like 0.71 or 0.6, but with 0.7, the final value is reported to be zero?

```
    template<
            int IntegerDigits,
            int FractionalDigits,
            class Narrowest>
    using safe_elastic_fixed_point = cnl::fixed_point<
            cnl::overflow_integer<
                    cnl::elastic_integer<
                            IntegerDigits+FractionalDigits,
                            Narrowest>,
                    cnl::throwing_overflow_tag>,
            -FractionalDigits>;


    using real_pos_t = safe_elastic_fixed_point<32, 8, std::int16_t>;
    real_pos_t not_zero = real_pos_t(7) / real_pos_t(127);
    ASSERT_NEAR(float(not_zero), 90.f/127.f, 0.01);
```

**Comment by johnmcfarlane** [comment]
tl;dr: The API has changed a little since the last time I updated the docs. (Sorry about that.) See option 1 below for the new API.

The result of the `/` operation has no fractional digits as both numerator and denominator have the same exponent. Here's [the same test](https://godbolt.org/z/zliRxc) without the conversion back to `real_pos_t`. It demonstrates that the result has no fractional digits.

*This is vexing* because the user almost always expects a result with the same precision as the inputs. However, consider `1/3`. How do you chose the right number of digits? The answer is that it's arbitrary and the user has to make a choice. The default which CNL provides is the one with the least run-time cost but there are some alternatives:

1. The free function, `divide`, has been replaced with a type, `fraction`. (It was briefly called [`fractional`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1050r0.html).) You can now chose exactly how much precision you get using the `fraction` API. Here's **[your example using `fraction`](https://godbolt.org/z/jb8XfQ)**.
2. You can manually increase the precision of the numerator ahead of performing the division. ([Example](https://godbolt.org/z/3qyINu))
3. Finally, for lossless arithmetic, you can deal with [quotient and remainder](https://godbolt.org/z/V-Ma-Y) just like with raw integer division. Remember that fixed-point is essentially a superset of integer and division is backed by integer division. Any attempts to make fixed-point division work like floating-point division will be limited.

(Note in that last example, I removed `elastic_integer` from the type. For some reason, it's producing an error. I've [filed an issue](https://github.com/johnmcfarlane/cnl/issues/325). It's probably not your first choice anyway but let me know if I'm wrong.)

I've also [filed an issue](https://github.com/johnmcfarlane/cnl/issues/324) to make it clear how to generate the docs in the README. The reason for the delay updating the docs is that a lot of source has moved around as I break down large header files and Doxygen isn't the best for diffing changes to make sure that nothing has been lost. Again, apologies for not doing this sooner.

**Comment by pgruenbacher** [comment]
this all worked for me thanks.

**Comment by johnmcfarlane** [comment]
CDSP is mentioned in the readme. And multiply is removed from the docs. However, there is still work to do with the examples on the front page.
